### ISSUE TITLE
Final Parameter

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/ModelicaSpecification#2035]
**Reported by patrick_marcel&#46;taeuber on 27 Jul 2016 11:41 UTC**
Please help me to understand the meaning of final parameters.

```
final parameter Real p1 = p2;
parameter Real p2 = 2;
```

How to handle the final attribute for p1?

a) p2 is implicitly marked as final and both parameters are evaluated during compilation to 2
b) both parameters are evaluated during simulation and the binding of p1 cannot be changed after compilation
c) something else

A similar case you can find in Spec: 9.3.1 Size Restriction on Connectors.

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/2035

**Comment by modelica-trac-importer** [comment]
**Comment by jmattsson on 28 Jul 2016 07:28 UTC**
I would say b. You should view it as the modification "= p2" being marked final rather than the parameter, and that expression can't change after compilation (but it's value can).

**Comment by modelica-trac-importer** [comment]
**Comment by choeger on 28 Jul 2016 09:08 UTC**
tl;dr: 
I agree with Jesper. 


The thing is, that Modelica is not a strictly evaluated language. Hence, you have to view `p1` as a method (in the OO-sense just without a mutable state). When you say that `p1` is "final", you prevent any modifications (overriding in the OO-lingo). An interpreter or compiler might make use of this by inlining the definition of `p1` wherever it is encountered.  

So yes, `p1` might be evaluated during simulation time. But it might also be replaced by an evaluation of `p2` - which in turn might be eliminated (consider an expression like `p1 - p2`) etc.

**Comment by modelica-trac-importer** [comment]
**Comment by patrick_marcel&#x2e;taeuber on 1 Aug 2016 09:22 UTC**
Thanks for your response.

So is it right to consider final parameters as protected parameters with the only difference that protecteds are not accessible for the user?



----

Seems that Dymola finalizes p1 even without the final attribute:

```
parameter Real p1 = p2;
parameter Real p2 = 2;
```
Only p2 can be modified after compilation.

Is this expected? What is then the additional effect of adding a final when all non constant bindings are finalized?

Same for this combination:

```
parameter Real p1 = p2;
final parameter Real p2 = 2;
```

I would expect that it should be possible to set p1 to any value after compilation because it is not finalized. But if non constant bindings are finalized (literally bound), none of both parameters can be changed after compilation.

In this case the final attribute only makes sense if the parameter has a constant binding value?
Is this interpretation correct?

Thanks in advance!

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 7 Sep 2016 09:46 UTC**
Replying to [comment:3 patrick_marcel.taeuber@…]:
> Thanks for your response.
> 
> So is it right to consider final parameters as protected parameters with the only difference that protecteds are not accessible for the user?

There is a difference: final means you cannot change it - protected that you cannot access it. 
Accessing parameters from other models is not that common, but possible:

```
model A
  final parameter Real p=1;
protected
  Real parameter q=2;
end A;
model B
  A a;
  parameter Real p=a.p; // Legal
  parameter Real q=a.q; // Illegal
end B;
```

> ----
> Seems that Dymola finalizes p1 even without the final attribute:
> 
> ```
> parameter Real p1 = p2;
> parameter Real p2 = 2;
> ```
> Only p2 can be modified after compilation.

Exactly what you can change after compilation isn't described in the specification.
So, there is no requirement that you can modify "p2".

> Is this expected? What is then the additional effect of adding a final when all non constant bindings are finalized?

Yes, it is expected.
The benefit of "final" is reuse inside models: think of large hierarchical model.

If you write `final parameter Real Area=Length^2;` inside a model you are preventing users from changing it to be anything but a square. It is also common to use final modifiers when propagating parameters (basically "final" means that the propagation cannot be removed):
```
  model A
    parameter Real p=1;
  end A;
  model B
    parameter Real p=1;
    A a1(final p=p);
    A a2(final p=p);
  end B;
```

I'm unsure if there is anything unclear in the specification related to this.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 30 Nov 2016 13:37 UTC**
As far as I can see this question about final parameters is now answered; and no change was needed.
